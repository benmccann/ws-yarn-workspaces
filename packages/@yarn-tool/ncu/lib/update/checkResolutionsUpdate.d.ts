/// <reference types="node" />
/// <reference types="bluebird" />
import { IPackageMap, IOptionsNpmCheckUpdates, IVersionCacheMapValue } from '../types';
import { EnumDetectYarnLock, IYarnLockSource } from '@yarn-tool/yarnlock-types';
export declare function checkResolutionsUpdate<T extends IYarnLockSource>(resolutions: IPackageMap, yarnlock_old_obj: IYarnLockSource | Buffer | string, options: Partial<IOptionsNpmCheckUpdates>): import("bluebird")<{
    verType: EnumDetectYarnLock.v1 | EnumDetectYarnLock.v2 | EnumDetectYarnLock.v3;
    yarnlock_old_obj: T;
    yarnlock_new_obj: T;
    update_list: string[];
    yarnlock_changed: boolean;
    deps: IVersionCacheMapValue[];
    deps2: IPackageMap<string[]>;
    deps3: Record<string, IVersionCacheMapValue>;
}>;
