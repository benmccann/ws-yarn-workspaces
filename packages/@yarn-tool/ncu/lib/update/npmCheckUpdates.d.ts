import { IOptionsNpmCheckUpdates } from '../types';
import { IWrapDedupeCache } from '@yarn-tool/yarnlock/lib/types';
import { ITSRequireAtLeastOne } from 'ts-type/index';
export declare function npmCheckUpdates<C extends IWrapDedupeCache>(cache: Partial<C>, ncuOptions: ITSRequireAtLeastOne<IOptionsNpmCheckUpdates, 'json_old' | 'packageData'>): Promise<Pick<IOptionsNpmCheckUpdates, "dep" | "minimal" | "newest" | "packageManager" | "registry" | "silent" | "greatest" | "upgrade" | "semverLevel" | "removeRange" | "dedupe" | "cwd" | "json_new" | "json_changed" | "list_updated" | "loglevel" | "versionTarget" | "current" | "noSafe"> & import("ts-type").ITSRequiredPick<IOptionsNpmCheckUpdates, "json_old"> & import("ts-type").ITSPartialPick<IOptionsNpmCheckUpdates, "packageData">>;
