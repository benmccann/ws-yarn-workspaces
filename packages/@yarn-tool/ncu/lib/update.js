"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.npmCheckUpdates = exports.checkResolutionsUpdate = void 0;
/**
 * Created by user on 2020/6/12.
 */
const npm_check_updates_1 = require("npm-check-updates");
const yarnlock_1 = require("@yarn-tool/yarnlock");
const bluebird_1 = __importDefault(require("bluebird"));
const util_1 = require("./util");
const semver_1 = __importDefault(require("semver"));
const remote_1 = require("./remote");
const options_1 = require("./options");
const table_1 = require("@yarn-tool/table");
const npm_package_arg_1 = __importDefault(require("npm-package-arg"));
const queryVersion_1 = require("@yarn-tool/pkg-version-query/lib/queryVersion");
const pkg_version_query_1 = require("@yarn-tool/pkg-version-query");
const parseSimpleSemVer_1 = require("@lazy-node/semver-simple-parse/lib/parseSimpleSemVer");
const mergeSimpleSemVer_1 = require("@lazy-node/semver-simple-parse/lib/mergeSimpleSemVer");
const stringifySimpleSemVer_1 = require("@lazy-node/semver-simple-parse/lib/stringifySimpleSemVer");
function checkResolutionsUpdate(resolutions, yarnlock_old_obj, options) {
    return bluebird_1.default.resolve()
        .then(async function () {
        if (typeof yarnlock_old_obj === 'string') {
            yarnlock_old_obj = yarnlock_1.parse(yarnlock_old_obj);
        }
        let result = yarnlock_1.filterResolutions({
            resolutions,
        }, yarnlock_old_obj);
        let deps = await remote_1.queryRemoteVersions(resolutions, options);
        //console.dir(deps);
        let deps2 = util_1.keyObjectToPackageMap(deps, true);
        let deps3 = Object.values(deps)
            .reduce(function (a, b) {
            a[b.name] = b;
            return a;
        }, {});
        let yarnlock_new_obj = {
            ...yarnlock_old_obj,
        };
        let update_list = [];
        let yarnlock_changed = false;
        Object.entries(result.max)
            .forEach(function ([name, data]) {
            let _key2 = name + '@' + deps3[name].version_old;
            /**
             * 檢查 版本範圍是否符合 與 版本是否不相同
             */
            //					console.dir({
            //						data,
            //						deps: deps2[name],
            //					});
            if (data.value.version != null && deps2[name] != null && semver_1.default.lt(data.value.version, deps2[name]) && yarnlock_new_obj[_key2] && yarnlock_new_obj[_key2].version != data.value.version) {
                Object.keys(result.deps[name])
                    .forEach(version => {
                    let key = name + '@' + version;
                    delete yarnlock_new_obj[key];
                });
                yarnlock_changed = true;
                update_list.push(name);
            }
            else {
                if (result.installed[name].length > 1) {
                    Object.keys(result.deps[name])
                        .forEach(version => {
                        let key = name + '@' + version;
                        yarnlock_new_obj[key] = data.value;
                    });
                    yarnlock_changed = true;
                }
            }
        });
        return {
            yarnlock_old_obj,
            yarnlock_new_obj,
            update_list,
            yarnlock_changed,
            deps,
            deps2,
            deps3,
        };
    });
}
exports.checkResolutionsUpdate = checkResolutionsUpdate;
async function npmCheckUpdates(cache, ncuOptions) {
    //ncuOptions.silent = false;
    //ncuOptions.json = false;
    //ncuOptions.cli = true;
    //ncuOptions.args = [];
    //ncuOptions.loglevel = 'verbose';
    ncuOptions = options_1.npmCheckUpdatesOptions(ncuOptions);
    ncuOptions.cwd = cache.cwd;
    ncuOptions.json_new = JSON.parse(ncuOptions.packageData);
    ncuOptions.list_updated = await npm_check_updates_1.run(ncuOptions);
    let json_changed = false;
    const current = {};
    const list_updated = {};
    await bluebird_1.default
        .resolve([
        'dependencies',
        'devDependencies',
        'peerDependencies',
        'optionalDependencies',
    ])
        .each(async (key) => {
        var _a;
        const deps = (_a = ncuOptions.json_new[key]) !== null && _a !== void 0 ? _a : {};
        await bluebird_1.default
            .resolve(Object.keys(deps))
            .each(async (name) => {
            var _a;
            const version_new = ncuOptions.list_updated[name];
            const version_old = deps[name];
            if (version_new === null || version_new === void 0 ? void 0 : version_new.length) {
                if (version_old !== version_new && util_1.allowUpdateVersion(version_old)) {
                    list_updated[name] = version_new;
                    current[name] = version_old;
                    deps[name] = version_new;
                    json_changed = true;
                }
            }
            else if (!/[\s|&]/.test(version_old)) {
                let key = `${name}@${version_old}`;
                let check = npm_package_arg_1.default(key);
                let prefix = (_a = /^([\^~\s]+)/.exec(version_old)) === null || _a === void 0 ? void 0 : _a[1];
                if ((prefix === null || prefix === void 0 ? void 0 : prefix.length) && check.type === 'range') {
                    let version_new = await queryVersion_1.queryVersionWithCache(name, version_old)
                        .then(v => prefix + v)
                        .catch(e => null);
                    if ((version_new === null || version_new === void 0 ? void 0 : version_new.length) && version_new !== version_old) {
                        try {
                            let { target } = mergeSimpleSemVer_1.mergeSimpleSemVer(parseSimpleSemVer_1.parseSimpleSemVer(version_old), parseSimpleSemVer_1.parseSimpleSemVer(version_new));
                            let version = stringifySimpleSemVer_1.stringifySemverFull(target);
                            if ((version === null || version === void 0 ? void 0 : version.length) > 0) {
                                list_updated[name] = version;
                                current[name] = version_old;
                                deps[name] = version;
                                json_changed = true;
                            }
                        }
                        catch (err) { }
                    }
                }
            }
        });
    });
    await pkg_version_query_1.getCache().fsDump();
    ncuOptions.json_changed = json_changed;
    ncuOptions.list_updated = list_updated;
    ncuOptions.current = current;
    const table = table_1.toDependencyTable({
        from: ncuOptions.current,
        to: ncuOptions.list_updated,
    }).toString();
    table && console.log(`\n${table}\n`);
    return ncuOptions;
}
exports.npmCheckUpdates = npmCheckUpdates;
//# sourceMappingURL=update.js.map