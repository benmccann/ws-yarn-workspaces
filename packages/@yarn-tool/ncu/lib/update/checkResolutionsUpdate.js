"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkResolutionsUpdate = void 0;
const bluebird_1 = __importDefault(require("bluebird"));
const util_1 = require("../util");
const semver_1 = __importDefault(require("semver"));
const parse_1 = require("@yarn-tool/yarnlock/lib/parse");
const core_1 = require("@yarn-tool/yarnlock/lib/core");
const queryRemoteVersions_1 = require("../remote/queryRemoteVersions");
function checkResolutionsUpdate(resolutions, yarnlock_old_obj, options) {
    return bluebird_1.default.resolve()
        .then(async function () {
        if (typeof yarnlock_old_obj === 'string') {
            yarnlock_old_obj = parse_1.parse(yarnlock_old_obj);
        }
        let result = core_1.filterResolutions({
            resolutions,
        }, yarnlock_old_obj);
        let deps = await queryRemoteVersions_1.queryRemoteVersions(resolutions, options);
        //console.dir(deps);
        let deps2 = util_1.keyObjectToPackageMap(deps, true);
        let deps3 = Object.values(deps)
            .reduce(function (a, b) {
            a[b.name] = b;
            return a;
        }, {});
        let yarnlock_new_obj = {
            ...yarnlock_old_obj,
        };
        let update_list = [];
        let yarnlock_changed = false;
        Object.entries(result.max)
            .forEach(function ([name, data]) {
            let _key2 = name + '@' + deps3[name].version_old;
            /**
             * 檢查 版本範圍是否符合 與 版本是否不相同
             */
            //					console.dir({
            //						data,
            //						deps: deps2[name],
            //					});
            if (data.value.version != null && deps2[name] != null && semver_1.default.lt(data.value.version, deps2[name]) && yarnlock_new_obj[_key2] && yarnlock_new_obj[_key2].version != data.value.version) {
                Object.keys(result.deps[name])
                    .forEach(version => {
                    let key = name + '@' + version;
                    delete yarnlock_new_obj[key];
                });
                yarnlock_changed = true;
                update_list.push(name);
            }
            else {
                if (result.installed[name].length > 1) {
                    Object.keys(result.deps[name])
                        .forEach(version => {
                        let key = name + '@' + version;
                        yarnlock_new_obj[key] = data.value;
                    });
                    yarnlock_changed = true;
                }
            }
        });
        return {
            yarnlock_old_obj,
            yarnlock_new_obj,
            update_list,
            yarnlock_changed,
            deps,
            deps2,
            deps3,
        };
    });
}
exports.checkResolutionsUpdate = checkResolutionsUpdate;
//# sourceMappingURL=checkResolutionsUpdate.js.map