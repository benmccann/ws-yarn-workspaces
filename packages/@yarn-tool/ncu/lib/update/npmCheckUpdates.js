"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.npmCheckUpdates = void 0;
const tslib_1 = require("tslib");
const options_1 = require("../options");
const npm_check_updates_1 = require("npm-check-updates");
const bluebird_1 = (0, tslib_1.__importDefault)(require("bluebird"));
const util_1 = require("../util");
const npm_package_arg_1 = (0, tslib_1.__importDefault)(require("npm-package-arg"));
const queryVersion_1 = require("@yarn-tool/pkg-version-query/lib/queryVersion");
const mergeSimpleSemVer_1 = require("@lazy-node/semver-simple-parse/lib/mergeSimpleSemVer");
const parseSimpleSemVer_1 = require("@lazy-node/semver-simple-parse/lib/parseSimpleSemVer");
const stringifySimpleSemVer_1 = require("@lazy-node/semver-simple-parse/lib/stringifySimpleSemVer");
const pkg_version_query_1 = require("@yarn-tool/pkg-version-query");
const deps_table_1 = (0, tslib_1.__importDefault)(require("@yarn-tool/table/lib/deps-table"));
async function npmCheckUpdates(cache, ncuOptions) {
    //ncuOptions.silent = false;
    //ncuOptions.json = false;
    //ncuOptions.cli = true;
    //ncuOptions.args = [];
    //ncuOptions.loglevel = 'verbose';
    ncuOptions = (0, options_1.npmCheckUpdatesOptions)(ncuOptions);
    ncuOptions.cwd = cache.cwd;
    ncuOptions.json_new = JSON.parse(ncuOptions.packageData);
    ncuOptions.list_updated = await (0, npm_check_updates_1.run)(ncuOptions);
    let json_changed = false;
    const current = {};
    const list_updated = {};
    await bluebird_1.default
        .resolve([
        'dependencies',
        'devDependencies',
        'peerDependencies',
        'optionalDependencies',
    ])
        .each(async (key) => {
        var _a;
        const deps = (_a = ncuOptions.json_new[key]) !== null && _a !== void 0 ? _a : {};
        await bluebird_1.default
            .resolve(Object.keys(deps))
            .each(async (name) => {
            var _a;
            const version_new = ncuOptions.list_updated[name];
            const version_old = deps[name];
            if (version_new === null || version_new === void 0 ? void 0 : version_new.length) {
                if (version_old !== version_new && (0, util_1.allowUpdateVersion)(version_old)) {
                    list_updated[name] = version_new;
                    current[name] = version_old;
                    deps[name] = version_new;
                    json_changed = true;
                }
            }
            else if (!/[\s|&]/.test(version_old)) {
                let key = `${name}@${version_old}`;
                let check = (0, npm_package_arg_1.default)(key);
                let prefix = (_a = /^([\^~\s]+)/.exec(version_old)) === null || _a === void 0 ? void 0 : _a[1];
                if ((prefix === null || prefix === void 0 ? void 0 : prefix.length) && check.type === 'range') {
                    let version_new = await (0, queryVersion_1.queryVersionWithCache)(name, version_old)
                        .then(v => prefix + v)
                        .catch(e => null);
                    if ((version_new === null || version_new === void 0 ? void 0 : version_new.length) && version_new !== version_old) {
                        try {
                            let { target } = (0, mergeSimpleSemVer_1.mergeSimpleSemVer)((0, parseSimpleSemVer_1.parseSimpleSemVer)(version_old), (0, parseSimpleSemVer_1.parseSimpleSemVer)(version_new));
                            let version = (0, stringifySimpleSemVer_1.stringifySemverFull)(target);
                            if ((version === null || version === void 0 ? void 0 : version.length) > 0) {
                                list_updated[name] = version;
                                current[name] = version_old;
                                deps[name] = version;
                                json_changed = true;
                            }
                        }
                        catch (err) { }
                    }
                }
            }
        });
    });
    await (0, pkg_version_query_1.getCache)().fsDump();
    ncuOptions.json_changed = json_changed;
    ncuOptions.list_updated = list_updated;
    ncuOptions.current = current;
    const table = (0, deps_table_1.default)({
        from: ncuOptions.current,
        to: ncuOptions.list_updated,
    }).toString();
    table && console.log(`\n${table}\n`);
    return ncuOptions;
}
exports.npmCheckUpdates = npmCheckUpdates;
//# sourceMappingURL=npmCheckUpdates.js.map