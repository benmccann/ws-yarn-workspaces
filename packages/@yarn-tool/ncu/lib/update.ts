/**
 * Created by user on 2020/6/12.
 */
import { run as _npmCheckUpdates } from 'npm-check-updates';
import {
	IPackageMap,
	IOptionsNpmCheckUpdates,
	IVersionCacheMapValue,
	IDependency,
	EnumVersionValue2,
	EnumVersionValue,
} from './types';
import {
	IYarnLockfileParseObject,
	parse as parseYarnLock,
	filterResolutions,
	IWrapDedupeCache,
} from '@yarn-tool/yarnlock';
import Bluebird from 'bluebird';
import { keyObjectToPackageMap, allowUpdateVersion } from './util';
import semver from 'semver';
import { queryRemoteVersions } from './remote';
import { npmCheckUpdatesOptions } from './options';
import { IPackageJsonDependenciesField } from '@ts-type/package-dts/package-json';
import { toDependencyTable } from '@yarn-tool/table';
import { ITSRequireAtLeastOne } from 'ts-type';
import npmPackageArg from 'npm-package-arg';
import { queryVersionWithCache } from '@yarn-tool/pkg-version-query/lib/queryVersion';
import { getCache } from '@yarn-tool/pkg-version-query';
import { parseSimpleSemVer } from '@lazy-node/semver-simple-parse/lib/parseSimpleSemVer';
import { mergeSimpleSemVer } from '@lazy-node/semver-simple-parse/lib/mergeSimpleSemVer';
import { stringifySemverFull } from '@lazy-node/semver-simple-parse/lib/stringifySimpleSemVer';

export function checkResolutionsUpdate(resolutions: IPackageMap,
	yarnlock_old_obj: IYarnLockfileParseObject | string,
	options: Partial<IOptionsNpmCheckUpdates>,
)
{
	return Bluebird.resolve()
		.then(async function ()
		{
			if (typeof yarnlock_old_obj === 'string')
			{
				yarnlock_old_obj = parseYarnLock(yarnlock_old_obj);
			}

			let result = filterResolutions({
				resolutions,
			}, yarnlock_old_obj);

			let deps = await queryRemoteVersions(resolutions, options);

			//console.dir(deps);

			let deps2 = keyObjectToPackageMap(deps, true);

			let deps3 = Object.values(deps)
				.reduce(function (a, b)
				{
					a[b.name] = b;

					return a;
				}, {} as Record<string, IVersionCacheMapValue>)
			;

			let yarnlock_new_obj: IYarnLockfileParseObject = {
				...yarnlock_old_obj,
			};

			let update_list: string[] = [];
			let yarnlock_changed = false;

			Object.entries(result.max)
				.forEach(function ([name, data])
				{
					let _key2 = name + '@' + deps3[name].version_old;

					/**
					 * 檢查 版本範圍是否符合 與 版本是否不相同
					 */
//					console.dir({
//						data,
//						deps: deps2[name],
//					});
					if (data.value.version != null && deps2[name] != null && semver.lt(data.value.version, deps2[name]) && yarnlock_new_obj[_key2] && yarnlock_new_obj[_key2].version != data.value.version)
					{
						Object.keys(result.deps[name])
							.forEach(version =>
							{
								let key = name + '@' + version;

								delete yarnlock_new_obj[key]
							})
						;

						yarnlock_changed = true;

						update_list.push(name);
					}
					else
					{
						if (result.installed[name].length > 1)
						{
							Object.keys(result.deps[name])
								.forEach(version =>
								{

									let key = name + '@' + version;

									yarnlock_new_obj[key] = data.value;
								})
							;

							yarnlock_changed = true;
						}
					}

				})
			;

			return {
				yarnlock_old_obj,
				yarnlock_new_obj,
				update_list,
				yarnlock_changed,
				deps,
				deps2,
				deps3,
			}
		})
		;
}

export async function npmCheckUpdates<C extends IWrapDedupeCache>(cache: Partial<C>,
	ncuOptions: ITSRequireAtLeastOne<IOptionsNpmCheckUpdates, 'json_old' | 'packageData'>,
)
{
	//ncuOptions.silent = false;

	//ncuOptions.json = false;
	//ncuOptions.cli = true;

	//ncuOptions.args = [];

	//ncuOptions.loglevel = 'verbose';

	ncuOptions = npmCheckUpdatesOptions(ncuOptions);

	ncuOptions.cwd = cache.cwd;

	ncuOptions.json_new = JSON.parse(ncuOptions.packageData);

	ncuOptions.list_updated = await _npmCheckUpdates(ncuOptions) as Record<string, string>;

	let json_changed = false;

	const current: IDependency = {};
	const list_updated: IDependency = {};

	await Bluebird
		.resolve([
			'dependencies',
			'devDependencies',
			'peerDependencies',
			'optionalDependencies',
		] as IPackageJsonDependenciesField[])
		.each(async (key) =>
		{
			const deps = ncuOptions.json_new[key] ?? {};

			await Bluebird
				.resolve(Object.keys(deps))
				.each(async (name) =>
				{
					const version_new = ncuOptions.list_updated[name];
					const version_old = deps[name];

					if (version_new?.length)
					{
						if (version_old !== version_new && allowUpdateVersion(version_old))
						{
							list_updated[name] = version_new;
							current[name] = version_old;

							deps[name] = version_new;

							json_changed = true;
						}
					}
					else if (!/[\s|&]/.test(version_old))
					{
						let key = `${name}@${version_old}`

						let check = npmPackageArg(key)
						let prefix = /^([\^~\s]+)/.exec(version_old)?.[1];

						if (prefix?.length && check.type === 'range')
						{
							let version_new = await queryVersionWithCache(name, version_old)
								.then(v => prefix + v)
								.catch(e => null)
							;

							if (version_new?.length && version_new !== version_old)
							{
								try
								{
									let { target } = mergeSimpleSemVer(parseSimpleSemVer(version_old), parseSimpleSemVer(version_new));

									let version = stringifySemverFull(target)

									if (version?.length > 0)
									{
										list_updated[name] = version;
										current[name] = version_old;

										deps[name] = version;

										json_changed = true;
									}
								}
								catch (err)
								{}
							}
						}
					}

				})

		})
	;

	await getCache().fsDump();

	ncuOptions.json_changed = json_changed;
	ncuOptions.list_updated = list_updated;
	ncuOptions.current = current;

	const table = toDependencyTable({
		from: ncuOptions.current,
		to: ncuOptions.list_updated,
	}).toString();

	table && console.log(`\n${table}\n`);

	return ncuOptions;
}
