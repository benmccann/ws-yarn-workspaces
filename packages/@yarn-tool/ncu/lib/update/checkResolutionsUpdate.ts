import { IPackageMap, IOptionsNpmCheckUpdates, IVersionCacheMapValue } from '../types';
import Bluebird from 'bluebird';
import { queryRemoteVersions } from '../remote';
import { keyObjectToPackageMap } from '../util';
import semver from 'semver';
import { IYarnLockfileParseObject } from '@yarn-tool/yarnlock/lib/types';
import {
	parse as parseYarnLock,
	filterResolutions,
	IWrapDedupeCache,
} from '@yarn-tool/yarnlock';

export function checkResolutionsUpdate(resolutions: IPackageMap,
	yarnlock_old_obj: IYarnLockfileParseObject | string,
	options: Partial<IOptionsNpmCheckUpdates>,
)
{
	return Bluebird.resolve()
		.then(async function ()
		{
			if (typeof yarnlock_old_obj === 'string')
			{
				yarnlock_old_obj = parseYarnLock(yarnlock_old_obj);
			}

			let result = filterResolutions({
				resolutions,
			}, yarnlock_old_obj);

			let deps = await queryRemoteVersions(resolutions, options);

			//console.dir(deps);

			let deps2 = keyObjectToPackageMap(deps, true);

			let deps3 = Object.values(deps)
				.reduce(function (a, b)
				{
					a[b.name] = b;

					return a;
				}, {} as Record<string, IVersionCacheMapValue>)
			;

			let yarnlock_new_obj: IYarnLockfileParseObject = {
				...yarnlock_old_obj,
			};

			let update_list: string[] = [];
			let yarnlock_changed = false;

			Object.entries(result.max)
				.forEach(function ([name, data])
				{
					let _key2 = name + '@' + deps3[name].version_old;

					/**
					 * 檢查 版本範圍是否符合 與 版本是否不相同
					 */
//					console.dir({
//						data,
//						deps: deps2[name],
//					});
					if (data.value.version != null && deps2[name] != null && semver.lt(data.value.version, deps2[name]) && yarnlock_new_obj[_key2] && yarnlock_new_obj[_key2].version != data.value.version)
					{
						Object.keys(result.deps[name])
							.forEach(version =>
							{
								let key = name + '@' + version;

								delete yarnlock_new_obj[key]
							})
						;

						yarnlock_changed = true;

						update_list.push(name);
					}
					else
					{
						if (result.installed[name].length > 1)
						{
							Object.keys(result.deps[name])
								.forEach(version =>
								{

									let key = name + '@' + version;

									yarnlock_new_obj[key] = data.value;
								})
							;

							yarnlock_changed = true;
						}
					}

				})
			;

			return {
				yarnlock_old_obj,
				yarnlock_new_obj,
				update_list,
				yarnlock_changed,
				deps,
				deps2,
				deps3,
			}
		})
		;
}
