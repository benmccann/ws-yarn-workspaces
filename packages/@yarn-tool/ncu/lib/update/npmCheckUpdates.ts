import { IOptionsNpmCheckUpdates, IDependency } from '../types';
import { npmCheckUpdatesOptions } from '../options';
import { run as _npmCheckUpdates } from 'npm-check-updates';
import Bluebird from 'bluebird';
import { IPackageJsonDependenciesField } from '@ts-type/package-dts/package-json';
import { allowUpdateVersion } from '../util';
import npmPackageArg from 'npm-package-arg';
import { queryVersionWithCache } from '@yarn-tool/pkg-version-query/lib/queryVersion';
import { mergeSimpleSemVer } from '@lazy-node/semver-simple-parse/lib/mergeSimpleSemVer';
import { parseSimpleSemVer } from '@lazy-node/semver-simple-parse/lib/parseSimpleSemVer';
import { stringifySemverFull } from '@lazy-node/semver-simple-parse/lib/stringifySimpleSemVer';
import { IWrapDedupeCache } from '@yarn-tool/yarnlock/lib/types';
import { ITSRequireAtLeastOne } from 'ts-type';
import { getCache } from '@yarn-tool/pkg-version-query';
import toDependencyTable from '@yarn-tool/table/lib/deps-table';

export async function npmCheckUpdates<C extends IWrapDedupeCache>(cache: Partial<C>,
	ncuOptions: ITSRequireAtLeastOne<IOptionsNpmCheckUpdates, 'json_old' | 'packageData'>,
)
{
	//ncuOptions.silent = false;

	//ncuOptions.json = false;
	//ncuOptions.cli = true;

	//ncuOptions.args = [];

	//ncuOptions.loglevel = 'verbose';

	ncuOptions = npmCheckUpdatesOptions(ncuOptions);

	ncuOptions.cwd = cache.cwd;

	ncuOptions.json_new = JSON.parse(ncuOptions.packageData);

	ncuOptions.list_updated = await _npmCheckUpdates(ncuOptions) as Record<string, string>;

	let json_changed = false;

	const current: IDependency = {};
	const list_updated: IDependency = {};

	await Bluebird
		.resolve([
			'dependencies',
			'devDependencies',
			'peerDependencies',
			'optionalDependencies',
		] as IPackageJsonDependenciesField[])
		.each(async (key) =>
		{
			const deps = ncuOptions.json_new[key] ?? {};

			await Bluebird
				.resolve(Object.keys(deps))
				.each(async (name) =>
				{
					const version_new = ncuOptions.list_updated[name];
					const version_old = deps[name];

					if (version_new?.length)
					{
						if (version_old !== version_new && allowUpdateVersion(version_old))
						{
							list_updated[name] = version_new;
							current[name] = version_old;

							deps[name] = version_new;

							json_changed = true;
						}
					}
					else if (!/[\s|&]/.test(version_old))
					{
						let key = `${name}@${version_old}`

						let check = npmPackageArg(key)
						let prefix = /^([\^~\s]+)/.exec(version_old)?.[1];

						if (prefix?.length && check.type === 'range')
						{
							let version_new = await queryVersionWithCache(name, version_old)
								.then(v => prefix + v)
								.catch(e => null)
							;

							if (version_new?.length && version_new !== version_old)
							{
								try
								{
									let { target } = mergeSimpleSemVer(parseSimpleSemVer(version_old), parseSimpleSemVer(version_new));

									let version = stringifySemverFull(target)

									if (version?.length > 0)
									{
										list_updated[name] = version;
										current[name] = version_old;

										deps[name] = version;

										json_changed = true;
									}
								}
								catch (err)
								{}
							}
						}
					}

				})

		})
	;

	await getCache().fsDump();

	ncuOptions.json_changed = json_changed;
	ncuOptions.list_updated = list_updated;
	ncuOptions.current = current;

	const table = toDependencyTable({
		from: ncuOptions.current,
		to: ncuOptions.list_updated,
	}).toString();

	table && console.log(`\n${table}\n`);

	return ncuOptions;
}
