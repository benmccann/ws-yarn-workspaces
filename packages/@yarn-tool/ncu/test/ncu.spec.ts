import IPackageJson from '@ts-type/package-dts/package-json';
import FastGlob from '@bluelovers/fast-glob';
import { basename, join } from 'path';
import { readJSONSync } from 'fs-extra';
import { npmCheckUpdates } from '../lib/update/npmCheckUpdates';

describe(`should not update`, () =>
{

	[
		">= 1",
		"> 1",
		"= 1",
		"< 1",
		"<= 1",
		"^1 | ^2",
		"^1 & ^2",
	].forEach(version => {

		const name = "@types/node";
		const field = "devDependencies";

		test(version, async () =>
		{

			let json_old: IPackageJson = {
				[field]: {
					[name]: version,
				},
			};

			let actual = await npmCheckUpdates({}, {
				json_old,
			})

			expect(actual.json_new[field]).toHaveProperty(name, version);
			expect(actual).toMatchSnapshot();


		});

	});

})

describe(`should update`, () =>
{

	FastGlob.sync([
		'*.json'
	], {
		cwd: join(__dirname, 'fixtures', 'update'),
		absolute: true,
	})
		.forEach(file => {

			test(basename(file), async () => {

				let json_old: IPackageJson = readJSONSync(file);

				let actual = await npmCheckUpdates({}, {
					json_old,
				})

				expect(actual.json_new)
					.not.toStrictEqual(json_old)
				;

			})

		})
	;

})
