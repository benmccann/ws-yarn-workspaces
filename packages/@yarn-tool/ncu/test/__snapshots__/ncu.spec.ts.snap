// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not update < 1 1`] = `
{
  "current": {},
  "cwd": undefined,
  "jsonUpgraded": true,
  "json_changed": false,
  "json_new": {
    "devDependencies": {
      "@types/node": "< 1",
    },
  },
  "json_old": {
    "devDependencies": {
      "@types/node": "< 1",
    },
  },
  "list_updated": {},
  "packageData": "{"devDependencies":{"@types/node":"< 1"}}",
  "packageManager": "npm",
}
`;

exports[`should not update <= 1 1`] = `
{
  "current": {},
  "cwd": undefined,
  "jsonUpgraded": true,
  "json_changed": false,
  "json_new": {
    "devDependencies": {
      "@types/node": "<= 1",
    },
  },
  "json_old": {
    "devDependencies": {
      "@types/node": "<= 1",
    },
  },
  "list_updated": {},
  "packageData": "{"devDependencies":{"@types/node":"<= 1"}}",
  "packageManager": "npm",
}
`;

exports[`should not update = 1 1`] = `
{
  "current": {},
  "cwd": undefined,
  "jsonUpgraded": true,
  "json_changed": false,
  "json_new": {
    "devDependencies": {
      "@types/node": "= 1",
    },
  },
  "json_old": {
    "devDependencies": {
      "@types/node": "= 1",
    },
  },
  "list_updated": {},
  "packageData": "{"devDependencies":{"@types/node":"= 1"}}",
  "packageManager": "npm",
}
`;

exports[`should not update > 1 1`] = `
{
  "current": {},
  "cwd": undefined,
  "jsonUpgraded": true,
  "json_changed": false,
  "json_new": {
    "devDependencies": {
      "@types/node": "> 1",
    },
  },
  "json_old": {
    "devDependencies": {
      "@types/node": "> 1",
    },
  },
  "list_updated": {},
  "packageData": "{"devDependencies":{"@types/node":"> 1"}}",
  "packageManager": "npm",
}
`;

exports[`should not update >= 1 1`] = `
{
  "current": {},
  "cwd": undefined,
  "jsonUpgraded": true,
  "json_changed": false,
  "json_new": {
    "devDependencies": {
      "@types/node": ">= 1",
    },
  },
  "json_old": {
    "devDependencies": {
      "@types/node": ">= 1",
    },
  },
  "list_updated": {},
  "packageData": "{"devDependencies":{"@types/node":">= 1"}}",
  "packageManager": "npm",
}
`;

exports[`should not update ^1 & ^2 1`] = `
{
  "current": {},
  "cwd": undefined,
  "jsonUpgraded": true,
  "json_changed": false,
  "json_new": {
    "devDependencies": {
      "@types/node": "^1 & ^2",
    },
  },
  "json_old": {
    "devDependencies": {
      "@types/node": "^1 & ^2",
    },
  },
  "list_updated": {},
  "packageData": "{"devDependencies":{"@types/node":"^1 & ^2"}}",
  "packageManager": "npm",
}
`;

exports[`should not update ^1 | ^2 1`] = `
{
  "current": {},
  "cwd": undefined,
  "jsonUpgraded": true,
  "json_changed": false,
  "json_new": {
    "devDependencies": {
      "@types/node": "^1 | ^2",
    },
  },
  "json_old": {
    "devDependencies": {
      "@types/node": "^1 | ^2",
    },
  },
  "list_updated": {},
  "packageData": "{"devDependencies":{"@types/node":"^1 | ^2"}}",
  "packageManager": "npm",
}
`;
